src/pip/_internal/req/__init__.py

```diff
@@ -53,40 +59,84 @@ def install_given_reqs(
             ', '.join([req.name for req in to_install]),
         )
 
-    installed = []
+    # pre allocate installed package names
+    installed = [None] * len(
+        to_install
+    )  # type: List[Union[None, InstallationResult, BaseException]]
+
+    install_args = [install_options, global_options, args, kwargs]
 
     with indent_log():
-        for requirement in to_install:
-            if requirement.should_reinstall:
-                logger.info('Attempting uninstall: %s', requirement.name)
-                with indent_log():
-                    uninstalled_pathset = requirement.uninstall(
-                        auto_confirm=True
-                    )
+        if to_install and Pool is not None:
+            # first let's try to install in parallel,
+            # if we fail we do it by order.
+            pool = Pool()
             try:
-                requirement.install(
-                    install_options,
-                    global_options,
-                    *args,
-                    **kwargs
-                )
-            except Exception:
-                should_rollback = (
-                    requirement.should_reinstall and
-                    not requirement.install_succeeded
+                pool_result = pool.map_async(
+                    partial(__single_install, install_args,
+                            in_subprocess=True), to_install
                 )
-                # if install did not succeed, rollback previous uninstall
-                if should_rollback:
-                    uninstalled_pathset.rollback()
+                # python 2.7 timeout=None will not catch KeyboardInterrupt
+                installed = pool_result.get(timeout=999999)
```

Comment:
I think `pool.imap[_unordered]` might be a better choice in this case.

Reply:
Nice, `Pool.imap_unordered` makes sense, I'll change
Regarding `freeze_support` that was just a test to see if it solves the Windows py2.7 issue, obviously it does not. on any other version it passed on Windows without `freeze_support`.
If you have an idea on why it fails on on Windows py2.7 , feel free to suggest )

Reply:
Regrading the inhering from Pool and adding capabilities, well it's a breeze to do, that said, I'm not sure the guys want to add another layer of utilities. For example I would not want to create another version of "dummy" Pool, just to make the code is "Pool agnostic", when we actually want accountability on when/how you use Pools. My 2 cents...

Reply:
Regarding the `imap` thing, what I meant was the `map_async` thing look cumbersome since we need to wait for it to finish anyway.  Either `[i]map` or `imap_unordered` is faster I don't know and I think the difference would be marginal if any.

> Regrading the inhering from Pool and adding capabilities

I don't think I get what you mean, because I did not mention inhering the Pool at all.  Just in case my guess is correct, you seem to refer to my comment on abstracting the call as a simple `map`-like function instead of *try Pool(); except for; else pool.map, pool.join, pool.close* which clutters the already complex logic.

Reply:
>what I meant was the map_async thing look cumbersome since we need to wait for it to finish anyway

The only reason for me to put it there is Python2.7 support.
For some reason with python2.7 if you do not use async & wait, it will not catch `ctrl-c` and will not break the process Pool, on higher python versions it was fixed, but in order to get to do act as expected I had to use async. The other option is to test for python version, but that seemed like something we should avoid...

>a simple map-like function instead of try Pool(); except for; else pool.map, pool.join, pool.close which clutters the already complex logic.

Sure, that sounds good.

Reply:
> For some reason with python2.7 if you do not use async & wait, it will not catch ctrl-c and will not break the process Pool

Thanks, TIL.  BTW seemingly we need to think about the way we handle exception here.  For example KeyboardInterrupt is sent to all processes in the pool, and thus should be handled by the function being fed to `map_async`.  AFAIK other exceptions raised from within the that function should look alright but I don't know for sure.

Reply:
Per you suggestion I extracted the Pool map method, in theory it could actually be reused.
As for the `KeyboardInterrupt`, we have to raise it in all the Pool sub-processes, otherwise the main process will not catch it, our mapped function actually takes care of that :)

Regrading other exceptions, actually I cannot think of any other than IO related, but for completes we catch them and pass the exception to the main process. This way the user is aware and the log is clean and will present the exceptions in the original order.